ndcube/mixins/ndcubeio.py

```diff
@@ -0,0 +1,236 @@
+
+    
+
+    # Here we create a read and write function to support I/O of NDCube files
+    # Currently read option is not supported
+    # Here we delegate the write method to save the files to different helper functions
+    # For different formats, we can define different helper functions, and based on the file type
+    # we can call them. Currently only FITS is supported.
+
+
+    def to_hdulist(self, hdu_mask='MASK', hdu_uncertainty='UNCERT'):
+        """Create a HDUList from a ImageHDU object
+
+        Parameters
+        ----------
+        hdu_mask, hdu_uncertainty : str, optional
+            If it is a string append this attribute to the HDUList as
+            `astropy.io.fits.ImageHDU` with the string as the extension name.
+            Default is `MASK` for hdu_mask, `UNCERT` for uncertainty and `None`
+            for flags.
+
+
+        Raises
+        ------
+        ValueError
+            - If `self._mask` is set but not a `numpy.ndarray`.
+            - If `self._uncertainty` is set but not a astropy uncertainty type
+            - If `self._uncertainty` is set but has another unit than `self.__data`.
+
+        Returns
+        -------
+        hdulist:
+            `astropy.io.fits.HDUList`
+        """
+        MISSING = 'MISNG{0}'
+        EXTRA_COORDS_LABEL = 'EXTCR{0}'
+
+        # ----------------------------------HDU0------------------------------
+        # Create a copy of the meta data to avoid changing of the header of data
+        if self.meta is not None:
+            header = fits.Header(self.meta.copy())
+
+        else:
+            header = fits.Header()
+        # Create a FITS Safe header from the meta data
+
+        # Cross-check if the metadata is fits safe, if not then it is corrected
+        for k, v in header.items():
+            _insert_in_metadata_fits_safe(header, k, v)
+
+        if self.wcs:
+            # Create a header for a given wcs object
+            # Hard-Coded relax parameter to write all
+            # recognized informal extensions of the WCS standard.
+            wcs_header = self.wcs.to_header(relax=True)
+            header.extend(wcs_header, useblanks=False, update=True)
+
+        # Create a FITS header for storing missing_axis
+        # MISSING0 : 1 if axis 1 is missing
+        # other keywords are skipped, if the axis is present
+        # The missing axis are expected to be present, as it is a bool
+        header0 = fits.Header()
+        for index, axis in enumerate(self.missing_axis):
+            if axis:
+                header0[MISSING.format(index)] = 1
+            else:
+                header0[MISSING.format(index)] = 0
+
+        for k, v in header0.items():
+            _insert_in_metadata_fits_safe(header0, k, v)
+
+        header.extend(header0, useblanks=False, update=True)
+
+        # PrimaryHDU list contains only meta, missing_axis and wcs as headers, no data
+        hdus = [fits.PrimaryHDU(data=None, header=header)]
```

Comment:
I generally don't like the pattern of splitting the metadata and the data up. If we are saving the array in HDU 2 then we should be saving the metadata there as well.

Reply:
This is the metadata is for the whole ```NDCube```.  It does not simply apply to the data array but also the uncertainty, mask, etc.  Therefore, I was thinking that the ```NDCube``` metadata dictionary and WCS values should be in its own HDU, while metadata specifically related to that data array, e.g. ```unit``` should be in the data HDU, while uncertainty-related metadata, e.g. uncertainty type, should be in the uncertainty HDU and extra coord metadata, e.g. axis, name, should be in the extra coord HDU, etc.